### PR TITLE
Fixed offloading for PyT version/ Added Attention activation offloading support/ Native FP8 support

### PR DESCRIPTION
**Pull request by sanandaraj5597**
PR does the following:

1) Offloading was breaking for the latest PyTorch versions because torch.compile tried to save some stray tensors which shouldn't be considered for offloading. Added that filter and fixed the issue.

2) Flash Attention and Fused attention can have it's input activations offloaded which wasn't supported before. Added support for that.

3) Native FP8 doesn't use transposed FP8 weights in a cache. Added None guards so that this doesn't break.

### CODE REVIEW

File: transformer_engine/pytorch/cpu_offload.py [hunk 1]
```diff
@@ -310,25 +320,33 @@ def get_tensor_buf_for_offloaded_tensor(self, tensor, tensor_tag):
...
-            # first copy the tensor to tensorbuf, so that the original tensor will not be deleted
-            tensor_buf = self.get_tensor_buf_for_offloaded_tensor(tensor, tensor_tag)
-            tensor_buf.copy_(tensor)
-            if hasattr(tensor,"weight_offloading"):
-                tensor_buf.weight_offloading = True
-            if hasattr(tensor,"activation_offloading"):
-                tensor_buf.activation_offloading = True
-           # Here we just save it, and at commit, bulk_offload_group will handle it
-            self.tensor_tag_to_state[tensor_tag] = tensor_buf
+        torch_stray_tensor = (type(tensor) != torch.Tensor) and check_torch_stray_tensor
```
Comment (sanandaraj5597) [inline transformer_engine/pytorch/cpu_offload.py:323]: When using torch.compile tensor types like FunctionalTensor and FakeTensors flow into this function which we want to avoid. Hence this check.

File: transformer_engine/pytorch/cpu_offload.py [hunk 2]
```diff
@@ -11,6 +11,15 @@
 
 __all__ = ['get_cpu_offload_context']
 
+TORCH_MAJOR = int(torch.__version__.split(".")[0])
```
[the remark this answers is not in the document]
Comment (ptrendx) [reply transformer_engine/pytorch/cpu_offload.py:14]: That said I'm not quite sure what is this exactly supposed to do? One could be using torch.compile in the earlier versions of pyTorch too instead of nvFuser, so this check seems wrong.

File: transformer_engine/pytorch/cpu_offload.py (hunk 1 above)
Comment (ptrendx) [reply transformer_engine/pytorch/cpu_offload.py:323]: Ok, but you are still saving them in line 347, so I'm a little confused here. Could you explain what exactly is this function supposed to do?

Comment (ptrendx) [reply transformer_engine/pytorch/cpu_offload.py:323]: Well, they don't need to be revived because you store the actual object so no need to revive anything?

File: transformer_engine/pytorch/cpu_offload.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanandaraj5597) [reply transformer_engine/pytorch/cpu_offload.py:14]: One reason could be, these tensors started popping in only after we started using NVFuser through torch.compile based on the version. 

Potentially we could check unconditionally, we just didn't have the need to in any of our NeMo/MCORE systems.
 
Also, even if stray tensors were there in the older version, they never went unread after saving. If we remove this check, we might miss the assert check where we ensure all tensors are read back after saving. Just didn't want some unknown errors in older version to get masked without my notice. Again, I'm leaving upto you, we can remove it if needed.

File: transformer_engine/pytorch/cpu_offload.py (hunk 1 above)
Comment (sanandaraj5597) [reply transformer_engine/pytorch/cpu_offload.py:323]: So what I mean by revive is autograd doing an implicit ctx.saved_tensors in backprop. For some reason (which only Torch developers know why), not all these Functional/Fake tensors are revived back during autograd, only a few are. So, assertion check fails for the one's that are not revived, hence this way of avoiding the check.

File: transformer_engine/pytorch/cpu_offload.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ptrendx) [reply transformer_engine/pytorch/cpu_offload.py:14]: I'd say we should remove it, but actually check for the expected types in the tensor_push function (so instead of checking if type is not Tensor, check if type is FakeTensor or FunctionalTensor.

[the remark this answers is not in the document]
Comment (sanandaraj5597) [reply transformer_engine/pytorch/cpu_offload.py]: @ptrendx I fixed as per your suggestion. Can we merge this if rest all are good?

**Review by ptrendx — APPROVED** [review]
